src/SAWScript/MGU.hs

```diff
@@ -67,44 +66,146 @@ listSubst = Subst . M.fromList
+ - a list of type pairs and printing them at the end. (That may have
+ - been a mistake; we'll see.)
+ -
+ - The last element (current function-type expected/found message) is
+ - always either a list of two message strings or empty. Function types
+ - we see go in it (replacing anything already there, so we keep only
+ - the outermost of a series) and are shifted out of it when we see
+ - something else. It could be a Maybe (String, String), but the code
+ - is noticeably more convenient the way it is.
+ -
+ - The initial message is kept separate so that the expected/found
+ - list can readily be built in either order. It's not clear if it's
+ - better to print the outermost or innermost mismatches first.
+ -
+ - Further notes on the message formatting:
+ -
+ - Print the expected and found types on their own lines. They can be
+ - large; if they are the resulting lines can still be fairly
+ - illegible, but at least the user doesn't have to hunt for "found"
+ - in the middle of a multi-line print.
+ -
+ - Pad the prefix of the prints so that the types line up; this is
+ - helpful for longer types that still fit on one output line.
+ -
+ - Indent each line with four spaces. What we send back gets printed
+ - underneath a message that's already (at least in some cases)
+ - indented by two spaces. It's important to make it clear that all
+ - the stuff we generate is part of that message and not, for example,
+ - an additional separate error.
+ -
+ - Note that although we append to the end of the expected/found list,
+ - we don't stick the start line in that list, because I keep going
+ - back and forth on whether the larger types should be printed first
+ - (prepending in failMGUadd) or last (appending). If we commit to
+ - appending we don't need to keep the start line separate.
+ -}
+
+data FailMGU = FailMGU String [String] [String]
+
+-- common code for printing expected/found types
+showTypes :: Type -> Type -> [String]
+showTypes ty1 ty2 =
+  let expected = "    Expected: " ++ pShow ty1
+      found    = "    Found:    " ++ pShow ty2
+  in
+  [expected, found]
+
+-- fail with expected/found types
+failMGU :: String -> Type -> Type -> Either FailMGU a
+failMGU start ty1 ty2 = Left (FailMGU start (showTypes ty1 ty2) [])
+
+-- fail with no types
+failMGU' :: String -> Either FailMGU a
+failMGU' start = Left (FailMGU start [] [])
+
+-- add another expected/found type pair to the failure
+-- (pull in the last function-type lines if any)
+failMGUAdd :: FailMGU -> Type -> Type -> FailMGU
+failMGUAdd (FailMGU start eflines lastfunlines) ty1 ty2 =
+  FailMGU start (eflines ++ lastfunlines ++ showTypes ty1 ty2) []
+
+-- add another pair that's a function type
+-- (overwrite any previous function type lines)
+failMGUAddFun :: FailMGU -> Type -> Type -> FailMGU
+failMGUAddFun (FailMGU start eflines _) ty1 ty2 =
+  FailMGU start eflines (showTypes ty1 ty2)
+
+-- unpack the failure into a string list for printing
+failMGUUnpack :: FailMGU -> [String]
```

Comment:
Given that this function is primarily meant for pretty-printing purposes, I would suggest naming this something like `ppFailMGU`, which is in line with the naming conventions we use for other pretty-printing functions.

Reply:
Fair enough. It started out life as a glorified string concat. Done in c718c106b1fcaa011b816a36ac12a0c84addd619